desc/equilibrium/equilibrium.py

```diff
@@ -836,60 +888,100 @@ def compute(
         if data is None:
             data = {}
 
-        # To avoid the issue of using the wrong grid for surface and volume averages,
-        # we first figure out what needed qtys are flux functions or volume integrals
-        # and compute those first on a full grid
         p = "desc.equilibrium.equilibrium.Equilibrium"
-        # If the user wants to compute x which depends on y which in turn depends on z,
-        # and they pass in y already computed in data, then we shouldn't need to compute
-        # z at all.
-        deps = list(
+        # TODO:
+        #  If the user wants to compute x which depends on y which in turn depends on z,
+        #  and they pass in y already in data, then we shouldn't need to compute z at
+        #  all. To resolve this we need to compute dependencies recursively, so that
+        #  priming data with just the first order dependencies will avoid computing
+        #  unnecessary dependencies. For now just subtract out y.
+        deps = (
             set(get_data_deps(names, obj=p, has_axis=grid.axis.size) + names)
-            - data.keys()  # subtract out y if already computed
+            - data.keys()
         )
-        # TODO: replace this logic with `grid_type` from data_index
-        dep0d = [
-            dep
-            for dep in deps
-            if (data_index[p][dep]["coordinates"] == "") and (dep not in data)
-        ]
-        dep1dr = [
+
+        def need_src(name):
+            # Need to compute these on grid that is paired to the source grid, since
+            # the compute logic assume input data is evaluated on those coordinates.
+            # We exclude these from the depXdx sets below since the grids we will
+            # use to compute those dependencies are coordinate-blind.
+            return bool(data_index[p][name]["source_grid_requirement"])
+
+        need_src_deps = _grow_seeds(set(filter(need_src, deps)), deps)
+
+        dep0d = {dep for dep in deps if self.is_0d(dep) and dep not in need_src_deps}
+        # Unless user asks, don't try to recompute stuff which are only dependencies
+        # of dep0d. Example, need R0. R0 <- A <- A(z) computable on dep0d grid.
+        # But A(z) in dep1dz and attempt to recompute on dep1dz grid will error
+        # without unique zeta idx defined, which mapped coordinate grids lack.
+        dep0d_deps = set(get_data_deps(dep0d, obj=p, has_axis=grid.axis.size))
+        # This filter is stronger than the name implies, but the false positives
+        # will still be computed correctly with the logic in compute.utils.compute.
+        just_dep0d_dep = lambda name: name in dep0d_deps and name not in names
+        dep1dr = {
             dep
             for dep in deps
-            if (data_index[p][dep]["coordinates"] == "r") and (dep not in data)
-        ]
-        dep1dz = [
+            if self.is_1dr(dep) and not just_dep0d_dep(dep) and dep not in need_src_deps
```

Comment:
This part of gridding is always a mess to understand so I am completely ignoring it.

Reply:
understandable

Reply:
can you explain the `just_dep0d_dep` part?

Reply:
I've lengthened the explanation in the comments in the code

Reply:
we wouldn't need crutches like filtering out just_dep0d_dep if we did just did the natural recursive algorithm